lib/src/rules/provide_correct_intl_args.dart

```diff
@@ -0,0 +1,252 @@
+        ?.firstOrDefault()
+        ?.as<StringInterpolation>();
+
+    if (messageArgument != null) {
+      final interpolationExpressions = messageArgument.elements
+          .whereType<InterpolationExpression>()
+          .map((expression) => expression.expression)
+          .toList();
+
+      addIssues(interpolationExpressions
+          .where((item) => item is! SimpleIdentifier)
+          .map((item) => _MustBeSimpleIdentifierIssue(item)));
+
+      final interpolationExpressionSimpleIdentifiers =
+          interpolationExpressions.whereType<SimpleIdentifier>().toList();
+
+      if (argsArgument != null &&
+          parameterSimpleIdentifiers.isEmpty &&
+          interpolationExpressionSimpleIdentifiers.isEmpty) {
+        addIssue(_ArgsMustBeOmittedIssue(argsArgument));
+      }
+
+      if (interpolationExpressionSimpleIdentifiers.isNotEmpty) {
+        _checkAllInterpolationMustBeContainsInParameters(
+            interpolationExpressionSimpleIdentifiers,
+            parameterSimpleIdentifiers);
+        _checkAllInterpolationMustBeContainsInArgs(
+            interpolationExpressionSimpleIdentifiers, argsSimpleIdentifiers);
+      }
+    } else {
+      for (final item in parameterSimpleIdentifiers) {
+        addIssue(_ParameterMustBeOmittedIssue(item));
+        addIssue(_ArgsItemMustBeOmittedIssue(item));
+      }
+    }
+  }
+
+  void _checkAllArgsElementsMustBeSimpleIdentifier(
+      ListLiteral argsListLiteral) {
+    final argsElements = argsListLiteral?.elements ?? <CollectionElement>[];
+
+    addIssues(argsElements
+        ?.where((item) => item is! SimpleIdentifier)
+        ?.map((item) => _MustBeSimpleIdentifierIssue(item)));
+  }
+
+  void _checkAllParametersMustBeContainsInArgs(
+      List<SimpleIdentifier> parameters, List<SimpleIdentifier> argsArgument) {
+    final argsNames = argsArgument.map((item) => item.token.value()).toSet();
+    addIssues(parameters
+        ?.where((param) => !argsNames.contains(param.token.value()))
+        ?.map((param) => _ParameterMustBeInArgsIssue(param)));
+  }
+
+  void _checkAllArgsMustBeContainsInParameters(
+      List<SimpleIdentifier> argsArgument, List<SimpleIdentifier> parameters) {
+    final parametersNames =
+        parameters.map((item) => item.token.value()).toSet();
+    addIssues(argsArgument
+        ?.where((arg) => !parametersNames.contains(arg.token.value()))
+        ?.map((arg) => _ArgsMustBeInParameterIssue(arg)));
+  }
+
+  void _checkAllInterpolationMustBeContainsInParameters(
+      List<SimpleIdentifier> simpleIdentifierExpressions,
+      List<SimpleIdentifier> parameters) {
+    final parametersNames =
+        parameters.map((item) => item.token.value()).toSet();
+    addIssues(simpleIdentifierExpressions
+        ?.where((param) => !parametersNames.contains(param.token.value()))
+        ?.map((param) => _InterpolationMustBeInParameterIssue(param)));
+  }
+
+  void _checkAllInterpolationMustBeContainsInArgs(
```

Comment:
It seems that all this four methods duplicate each other except used class instance. Should we really use multiple classes instead of plain strings with one class then?

Reply:
A bit of unified code, I think the four methods provide a better understanding of the logic